src/number_theory/legendre_symbol/gauss_sum.lean

```diff
@@ -0,0 +1,418 @@
+* `char.card_pow_card`: When `F` and `F'` are finite fields and `χ : F → F'`
+  is a nontrivial quadratic character, then `(χ (-1) * #F)^(#F'/2) = χ (#F')`.
+* `char.two_pow_card`: For every finite field `F` of odd characteristic,
+  we have `2^(#F/2) = χ₈(#F)` in `F`.
+
+This machinery can be used to derive (a generaliation of) the Law of
+Quadratic Reciprocity.
+
+## Tags
+
+additive character, multiplicative character, Gauss sum
+-/
+
+namespace char
+
+universes u v
+
+open_locale big_operators
+
+open add_char mul_char multiplicative
+
+section gauss_sum_def
+
+-- `R` is the domain of the characters
+variables {R : Type u} [comm_ring R] [fintype R]
+-- `R'` is the target of the characters
+variables {R' : Type v} [comm_ring R']
+
+/-!
+### Definition and first properties
+-/
+
+/-- Definition of the Gauss sum associated to a multiplicative and an additive character. -/
+def gauss_sum (χ : mul_char R R') (ψ : add_char R R') : R' := ∑ a, χ a * ψ (of_add a)
+
+/-- Replacing `ψ` by `mul_shift ψ a` and multiplying the Gauss sum by `χ a` does not change it. -/
+lemma gauss_sum_mul_shift (χ : mul_char R R') (ψ : add_char R R') (a : Rˣ) :
+  χ a * gauss_sum χ (mul_shift ψ a) = gauss_sum χ ψ :=
+begin
+  simp only [gauss_sum, mul_shift, monoid_hom.coe_comp, function.comp_app, to_add_of_add,
+             add_monoid_hom.to_multiplicative_apply_apply, add_monoid_hom.coe_mul_left],
+  rw [finset.mul_sum],
+  conv in (_ * (_ * _)) {rw [← mul_assoc, ← map_mul]},
+  exact fintype.sum_bijective _ (units.mul_left_bijective a) _ _ (λ x, rfl),
+end
+
+end gauss_sum_def
+
+/-!
+### The product of two Gauss sums
+-/
+
+section gauss_sum_prod
+
+-- In the following, we need `R` to be a finite field and `R'` to be a domain.
+variables {R : Type u} [field R] [fintype R] {R' : Type v} [comm_ring R'] [is_domain R']
+
+-- Two helper lemmas for `gauss_sum_mul_gauss_sum_eq_card` below
+private
+lemma gauss_sum_mul_aux₁ {χ : mul_char R R'} (hχ : is_nontrivial χ) (ψ : add_char R R') (b : R) :
+  ∑ a, χ (a * b⁻¹) * ψ (of_add (a - b)) = ∑ c, χ c * ψ (of_add (b * (c - 1))) :=
+begin
+  cases eq_or_ne b 0 with hb hb,
+  { -- case `b = 0`
+    simp only [hb, inv_zero, mul_zero, mul_char.map_zero, zero_mul, finset.sum_const_zero,
+               of_add_zero, monoid_hom.map_one, mul_one],
+    exact hχ.sum_eq_zero.symm, },
+  { -- case `b ≠ 0`
+    symmetry,
+    let lhs := λ c, χ c * ψ (of_add (b * (c - 1))),
+    let rhs := λ a, χ (a * b⁻¹) * ψ (of_add (a - b)),
+    have lr : ∀ x : R, lhs x = rhs (b * x) :=
+    begin
+      intro x,
+      simp only [lhs, rhs],
+      rw [mul_assoc, mul_comm x, ← mul_assoc, mul_inv_cancel hb, one_mul, mul_sub, mul_one],
+    end,
+    exact fintype.sum_bijective _ (mul_left_bijective₀ b hb) lhs rhs lr, },
+end
+
+private
+lemma gauss_sum_mul_aux₂ [decidable_eq R] (ψ : add_char R R') (b : R) (hψ : is_primitive ψ) :
+  ∑ (x : R), ψ (of_add (x * (b - 1))) = if b = 1 then fintype.card R else 0 :=
+begin
+  split_ifs,
```

Comment:
Please name new assumptions explicitly if you refer to them later on.
```suggestion
  split_ifs with h,
```

Reply:
OK.